File: 5f9aba12-2f21-4ed8-8aa6-9283a9ab9fe0/src/main/java/com/sportsevent/repository/NotificationRepository.java
package com.sportsevent.repository;

import com.sportsevent.entity.Notification;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;
import org.springframework.stereotype.Repository;

import java.time.LocalDateTime;
import java.util.List;

@Repository
public interface NotificationRepository extends MongoRepository<Notification, String> {

    List<Notification> findByStatus(Notification.NotificationStatus status);

    @Query("{'recipients.recipientId': ?0}")
    List<Notification> findByRecipientId(String recipientId);

    @Query("{'relatedEntityId': ?0, 'relatedEntityType': ?1}")
    List<Notification> findByRelatedEntity(String relatedEntityId, String relatedEntityType);

    @Query("{'status': {$in: ['PENDING', 'FAILED']}, 'scheduledAt': {$lte: ?0}}")
    List<Notification> findPendingNotifications(LocalDateTime now);
}

File: 9096d8e2-d12b-468a-91ae-53c97e7bf587/backend/src/main/kotlin/com/ems/dispatch/repository/NotificationRepository.kt
package com.ems.dispatch.repository

import com.ems.dispatch.entity.Notification
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.time.LocalDateTime

@Repository
interface NotificationRepository : JpaRepository<Notification, Long> {
    fun findByNotificationNo(notificationNo: String): Notification?
    fun findByRecipientUserIdOrderByCreatedAtDesc(recipientUserId: Long): List<Notification>
    fun findByRecipientHospitalIdOrderByCreatedAtDesc(recipientHospitalId: Long): List<Notification>
    fun findByStatus(status: String): List<Notification>
    fun findByDispatchEventId(dispatchEventId: Long): List<Notification>

    @Query(
        """
        SELECT n FROM Notification n 
        WHERE n.recipientUser.id = :userId 
        AND n.status IN :statuses 
        ORDER BY n.createdAt DESC
        """
    )
    fun findByUserIdAndStatuses(
        @Param("userId") userId: Long,
        @Param("statuses") statuses: List<String>
    ): List<Notification>

    @Query(
        """
        SELECT COUNT(n) FROM Notification n 
        WHERE n.recipientUser.id = :userId 
        AND n.status = 'DELIVERED'
        AND n.readAt IS NULL
        """
    )
    fun countUnreadByUserId(@Param("userId") userId: Long): Long

    fun deleteByCreatedAtBefore(timestamp: LocalDateTime)
}
